### PR TITLE
improve form validation

### LINKED ISSUE #20: Form Validation
- [x] The website field should validate that the input is a URL
- [ ] All fields should be alphanumeric (? see below for discussion)

### PR DESCRIPTION
**Pull request by PsychicNoodles**
## Resolves #20

Updated form validation to use Vuelidate and added a valid characters validation string (on the i18n repo: https://github.com/ourjapanlife/findadoc-localization/pull/60).

### How to test

Try inputting something invalid on the add clinic page. The current validations are:

* city
  * required
  * minimum length of 2
  * valid characters
* ward
  * required
  * minimum length of 2
  * valid characters
* clinic (previously called `name`)
  * required
  * minimum length of 2
  * valid characters
* website
  * required
  * url (this covers valid characters)

### Screenshots

https://user-images.githubusercontent.com/1944082/123417099-516f9080-d5f2-11eb-808f-34317a19e950.mp4

### CODE REVIEW

File: nuxt.config.js
Comment (Ermish) [inline nuxt.config.js:84]: I don't think the i18n files are supposed to be included in the MR. 

@ann-kilzer or @theyokohamalife can you confirm?

File: pages/add-clinic.vue
Comment (Ermish) [inline pages/add-clinic.vue:115]: is there a `$valid` option vs `!$invalid`?

File: pages/add-clinic.vue
Comment (Ermish) [inline pages/add-clinic.vue:140]: if you invert this it will be easier to read. You won't have to nest the whole block of logic.

File: pages/add-clinic.vue
Comment (Ermish) [inline pages/add-clinic.vue:141]: Can we break this up to be more readable?
something like:

```
    const whatYourFilteringHere = field.$flattenParams().filter((param) => !field[param.name])
    const whatTheMapFunctionIsReturning = whatYourFilteringHere
    
    return whatTheMapFunctionIsReturning 
```

It's a good practice to put your return on a separate line if it's not clear what is being returned. 
(something like `return 25 + x` is easy to read)

File: pages/add-clinic.vue
Comment (Ermish) [inline pages/add-clinic.vue:145]: should explain what `args` these are. `translationPosition`?

(I don't actually know what the second param of `$t` does)

File: pages/add-clinic.vue
Comment (Ermish) [reply pages/add-clinic.vue:140]: you can also get rid of the `else` when you do this

**Review by Ermish — CHANGES_REQUESTED** [review]
left a few comments. 
Thanks for helping out!!

File: pages/add-clinic.vue
Comment (PsychicNoodles) [reply pages/add-clinic.vue:115]: Unfortunately not, that'd make things easier to parse...
https://vuelidate.js.org/#sub-v-values

File: pages/add-clinic.vue
Comment (PsychicNoodles) [reply pages/add-clinic.vue:140]: Inverted and changed the condition to `field.$error`

File: pages/add-clinic.vue
Comment (PsychicNoodles) [reply pages/add-clinic.vue:141]: Split up the filter and map lines.

File: pages/add-clinic.vue
Comment (PsychicNoodles) [reply pages/add-clinic.vue:145]: Renamed it to `templateArgs` -- it's for plugging into variables into the template strings.

**Review by Ermish — APPROVED** [review]
Nice work! LGTM!

**Review by theyokohamalife — CHANGES_REQUESTED** [review]
Thank you!! It looks good but please remove the i18n files so there aren't merge conflicts.

**Comment by PsychicNoodles** [comment]
> Thank you!! It looks good but please remove the i18n files so there aren't merge conflicts.

@theyokohamalife The error display will probably break if this is merged without the i18n changes, so should we get that PR merged first and then pull in the latest i18n changes in this PR?

**Comment by Ermish** [comment]
> > Thank you!! It looks good but please remove the i18n files so there aren't merge conflicts.
> 
> @theyokohamalife The error display will probably break if this is merged without the i18n changes, so should we get that PR merged first and then pull in the latest i18n changes in this PR?

@PsychicNoodles I think we talked about this in slack, but I would recommend merging in the i18n PR, then updating / merging this one. This should ready to merge when that's done!

File: nuxt.config.js
Comment (ann-kilzer) [reply nuxt.config.js:84]: It's okay to bump them in the PR only if there were corresponding changes, but we should update them to the latest before merging.

File: package.json
Comment (ann-kilzer) [inline package.json:21]: 👍️

File: nuxt.config.js
Comment (ann-kilzer) [inline nuxt.config.js:87]: It's a little odd that it's referring to something under the homedir, no? Do we need to change the setup steps? 🤔

**Comment by ann-kilzer** [comment]
I'd be eager to get this in with the number of junk submissions coming in. Could you please fix the merge conflict in the i18n repo, fix the pages/add-clinic.vue conflict here, and I can help fix the i18n submodule?

Pardon the challenges with submodules, I am working on making the process smoother.

Thanks again for your work here

**Comment by PsychicNoodles** [comment]
Sure, I'll get on it as soon as I'm back home!

File: nuxt.config.js
Comment (PsychicNoodles) [reply nuxt.config.js:87]: I thought it was odd too, but I think the `~` here is actually just the project root? It's what the nuxt docs suggest here:
https://nuxtjs.org/docs/2.x/directory-structure/plugins

Comment (ann-kilzer) [reply nuxt.config.js:87]: Yeah i was googling it and it seems that most of the documentation follows this pattern. Let's try it  ¯\_(ツ)_/¯

**Comment by ann-kilzer** [comment]
Great, I approved the localization code, so to bring it up to date, please run the following:

```sh
cd i18n
git checkout main
git pull
cd ..
git add i18n
git commit -m "bump submodule"
git push
```

And we should be in business 👯

**Comment by ann-kilzer** [comment]
Almost there, don't forget to resolve the conflicts @PsychicNoodles 🙇🏻‍♀️

**Comment by PsychicNoodles** [comment]
Alright, everything should be fixed up now!

**Review by theyokohamalife — APPROVED** [review]
LGTM! Thank you!!

**Comment by ann-kilzer** [comment]
COVID-19_Vaccination_Info


Excellent

**Review by ann-kilzer — APPROVED** [review]
Thank you

**Pull request merged**
